### ISSUE TITLE
/controlled-inputs-using-react-hooks/

### ISSUE THREAD

**Issue by panzerdp** [panzerdp/dmitripavlutin.com-comments#85]
Written on 09/26/2020 17:36:25 

URL: https://dmitripavlutin.com/controlled-inputs-using-react-hooks/

**Comment by panzerdp** [comment]
> Comment written by **Alexandre Dy La Fuente** on **09/29/2020 14:19:52**

<p>This is an amazing explanation. I prefer component controlled too. Thanks for more one great post.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **09/29/2020 14:35:51**

<p>You're welcome Alexandre.</p>

**Comment by panzerdp** [comment]
> Comment written by **Bohdan Zamchenco** on **10/03/2020 10:20:52**

<p>I've never seen before value debouncing:) Usually, a more common approach is to debounce the onChange function. Thanks for this!</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **10/10/2020 07:10:09**

<p>You're welcome.</p>

**Comment by JamesTheBald** [comment]
Great tutorials. A quick note: the codesandbox examples for sections 3 and 4 of this tutorial only worked for me once I removed "toLowerCase()." from the filteredEmployees function. (I'm running Firefox on a Mac but I don't think that should matter.)

**Comment by panzerdp** [comment]
> Great tutorials. A quick note: the codesandbox examples for sections 3 and 4 of this tutorial only worked for me once I removed "toLowerCase()." from the filteredEmployees function. (I'm running Firefox on a Mac but I don't think that should matter.)

Thanks for letting me know @JamesTheBald. I lower cased the query too, and now it works correctly.

**Comment by HugaidaS** [comment]
Thanks for the great article, I have never seen this approach to controlled conponents before. I think more junior developer should read this!

**Comment by panzerdp** [comment]
> Thanks for the great article, I have never seen this approach to controlled conponents before. I think more junior developer should read this!

Glad you like it @Ellinsa!

**Comment by jaderz** [comment]
Hi.. great tutorial. Thank you! I'm new to React. Wondering if there is an easy way to then take your input values say on just a plain text input and then store those values into a variable or object to call for use in a different element? So if the first input had a value of 1 entered and second input had 2 entered, I'm looking to store the input data, then call a variable (or other type of element) that will display the input data 12 by just listing out the values very plain.

**Comment by panzerdp** [comment]
> Hi.. great tutorial. Thank you! I'm new to React. Wondering if there is an easy way to then take your input values say on just a plain text input and then store those values into a variable or object to call for use in a different element? So if the first input had a value of 1 entered and second input had 2 entered, I'm looking to store the input data, then call a variable (or other type of element) that will display the input data 12 by just listing out the values very plain.

Just use 2 state variables to control the 2 inputs.

**Comment by verlok** [comment]
I'm trying to create a script to run in my browser console to fill a form with some data that I provide. Something like:

```js
emailField.focus();
emailField.value = email;
emailField.blur();
// same for other fields
```

The thing is, when I do that and submit the form, the application tells me I still need to fill up some required fields. 
This is likely because the application was made in React and using controlled components, so the form correctness is checked against the form component state and not on what's the actual value of the form fields.

How do I make react copy in its state the values I've filled?

I've already tried:

```js
emailField.dispatchEvent(new Event('change'));
```

I've also tried

```js
emailField.dispatchEvent(new KeyboardEvent('keypress',{'key':'a'}));
```

Bot they didn't work. I keep getting the "please fill all the required fields" message.

Any help?

**Comment by panzerdp** [comment]
@verlok Try to check this [answer](https://stackoverflow.com/a/50219991/1894471), maybe it could be helpful.

**Comment by verlok** [comment]
> @verlok Try to check this [answer](https://stackoverflow.com/a/50219991/1894471), maybe it could be helpful.

It’s useful to know that the browser ignores keyboard events to populate inputs. Then how could I do that? Do you think I need puppeteer or nightwatch JS to command a browser with Node?

**Comment by panzerdp** [comment]
> > @verlok Try to check this [answer](https://stackoverflow.com/a/50219991/1894471), maybe it could be helpful.
> 
> It’s useful to know that the browser ignores keyboard events to populate inputs. Then how could I do that? Do you think I need puppeteer or nightwatch JS to command a browser with Node?

Yes, you could try to use puppeteer or nightwatch.

**Comment by amandaolsen** [comment]
This is a wonderful explanation! Thank you!!

Question: given that what makes a component controlled is whether state stores the form value, does a developer actually have to do the following to have created a controlled component?
`onChange = event => setVal(event.target.value)` and then attach the handler

Couldn't a developer add the following to the input field and still have a controlled component?
`onChange={(e) => setValue(e.target.value)}`

**Comment by panzerdp** [comment]
> This is a wonderful explanation! Thank you!!
> 
> Question: given that what makes a component controlled is whether state stores the form value, does a developer actually have to do the following to have created a controlled component? `onChange = event => setVal(event.target.value)` and then attach the handler
> 
> Couldn't a developer add the following to the input field and still have a controlled component? `onChange={(e) => setValue(e.target.value)}`

Yes, in both cases you have a controlled component, given that you also have `<input values={value} ... />`.

**Comment by gopeshkh1** [comment]
Shouldn't we setTimeout only when there is no id?

**Comment by kuyfer** [comment]
thanks
